### PR TITLE
Allow customizing the floating buttons

### LINKED ISSUE #2749: Allow the application to curate and customize floating map buttons
We’ve had a number of requests over the years to make the floating map buttons more configurable – to remove a button (#2102), add a button, change the look of a button (#2103), move the buttons to the other side of NavigationViewController, lay them out horizontally (#1139), and so on. It’s already possible to remove the feedback button by setting `NavigationViewController.showsReportFeedback`, but the other buttons are hard-coded.

CarPlayManager already has a hook for customizing the set of map buttons in CarPlay, via the pass-through `carPlayManager(_:mapButtonsCompatibleWith:in:for:)` method. A similar method on NavigationViewControllerDelegate would be beneficial for the main screen. We could also adapt the `MGLOrnamentPosition` approach in mapbox/mapbox-gl-native#13911 to make it easy to move the stack of buttons around.

/cc @mapbox/navigation-ios

### PR DESCRIPTION
**Pull request by ShanMa1991**
### Description

This pr is to add the ability to customize the floating buttons of navigation view, including adding and removing buttons from the default setting, changing the layout and position of the floating buttons.

- [x] fix #2749 
- [x] update change log
- [x] allow the floating buttons to change position between `topLeading` and `topTrailing`
- [x] allow the floating buttons to add and remove button

### Implementation

By creating the property of `floatingButtons`  and the `floatingButtonsPosition` in `NavigationView` and `RouteMapViewController`, it allows the customers to reach and edit the  `floatingButtons` through the `NavigationViewController`.

Create an enum of `MapOrnamentPosition` to control the position of speed limit view and floating buttons. It allows the customer to switch the position of the floating buttons between `topLeading` and `topTrailing`. The `NavigationViewController.floatingButtonsPosition` is default as `MapOrnamentPosition.topTrailing`.

 The `NavigationViewController.floatingButtons` is an array of `UIButton ` that supports to add or remove customized buttons. The default buttons are `FloatingButton`, which is declared with `FloatingButton.rounded(image:selectedImage:size:)` to be consistent.

For example, in the GuidanceCard UI,  by setting the following property through the navigation view controller, we could et the position of the floating buttons to at the top left, keep the middle button from original view setting and add a red button below.
```
        navigationViewController.floatingButtonsPosition = .topLeading
        let tempButton = Button(frame: CGRect(x: 100, y: 100, width: 50, height: 50))
        tempButton.backgroundColor = UIColor.red
        if let tempFloatingButton = navigationViewController.floatingButtons?[1]{
            navigationViewController.floatingButtons = [tempFloatingButton, tempButton]
        }
```
CustomizedFloatingButtons

fix #2749  
/cc @mapbox/navigation-ios

### CODE REVIEW

File: MapboxNavigation/NavigationView.swift
Comment (1ec5) [inline MapboxNavigation/NavigationView.swift:173]: `Dictionary.values` isn’t guaranteed to be in any particular order. In the screenshot above, for example, the mute button appears below the feedback button, opposite of the default order on main. Consider using an array to store the buttons instead of a dictionary. If you need to associate each button with an ID of some sort, there are multiple possible approaches, such as making `floatingButtons` a dictionary of tuples or making the IDs properties of the respective button classes.

File: MapboxNavigation/NavigationViewController.swift
Comment (1ec5) [inline MapboxNavigation/NavigationViewController.swift:214]: This design requires each button to have a unique ID. In general, UIKit and SwiftUI operate under the assumption that views do not have unique IDs. Instead, application code normally holds onto pointers to specific view instances in memory. In other words, if the application wants to specify a mute button, it should refer to `NavigationView.muteButton` (though that property would need to be exposed publicly through NavigationViewController).

File: MapboxNavigation/NavigationView.swift
Comment (ShanMa1991) [reply MapboxNavigation/NavigationView.swift:173]: The sequence of the dictionary could be fixed by replacing the `Array(Dictionary.values)` with `Dictionary.map{$0.value}`. But the problem of this design has some other problem related with the key value assigning as abobe. I would go back to the button arrays instead of the hashMap. Thanks for the suggestion!

File: MapboxNavigation/NavigationViewController.swift
Comment (ShanMa1991) [reply MapboxNavigation/NavigationViewController.swift]: Yes. The dictionary design has some problems related with the pointing. The array could allow the button to change the image and  the other property but still keeps its function. For example, by replacing the image and change the color, but this button still keeps the function of `mute` with pointing to the original `NavigationView.muteButton`  as the gif below:

```
        let tempButton = Button(frame: CGRect(x: 100, y: 100, width: 50, height: 50))
        tempButton.backgroundColor = UIColor.red
        if let tempFloatingButton = navigationViewController.floatingButtons?[1],
           let temp2 = navigationViewController.floatingButtons?[0] {
            tempFloatingButton.backgroundColor = UIColor.green
            tempFloatingButton.setImage(temp2.currentImage, for: .normal)
            navigationViewController.floatingButtons = [tempFloatingButton, tempButton]
        }
```
customizedMute

File: MapboxNavigation/NavigationViewController.swift
Comment (1ec5) [inline MapboxNavigation/NavigationViewController.swift:203]: Instead of declaring the enumeration as a nested type and exposing it at the top level via a type alias, declare the enumeration here at the top level, so no type alias is needed.

File: MapboxNavigation/NavigationView.swift
Comment (1ec5) [inline MapboxNavigation/NavigationView.swift:78]: For consistency with the map SDK’s `MGLOrnamentPosition` enumeration, consider calling this enumeration `MapOrnamentPosition` (singular). We’ll probably eventually need to use this same enumeration for other kinds of controls that are positioned over the map. For example, eventually, we may add the ability to place the floating buttons along the bottom or place the speed limit view on the same side as the floating buttons.

File: MapboxNavigation/NavigationView.swift
Comment (1ec5) [inline MapboxNavigation/NavigationView.swift:79]: The enumeration and its cases should each be documented to ensure that they appear in the jazzy docset.

File: MapboxNavigation/NavigationViewLayout.swift
Comment (1ec5) [inline MapboxNavigation/NavigationViewLayout.swift:34]: Use `.topRight` here so that, if we ever add more cases to the enumeration, this switch statement will generate a warning about exhaustiveness so we know to update it.

File: MapboxNavigation/NavigationView.swift
Comment (1ec5) [inline MapboxNavigation/NavigationView.swift:79]: This PR uses `leading` and `trailing` constraints, which means the sides are swapped in right-to-left languages. To avoid confusion, consider calling the cases `topLeading` and `topTrailing`, which would be consistent with `NSRectAlignment`. (Ideally we’d just use `NSRectAlignment`, but that enumeration is only available on iOS 13 and above.)

File: MapboxNavigation/NavigationViewController.swift
Comment (1ec5) [inline MapboxNavigation/NavigationViewController.swift:205]: Remember to add documentation about these two public properties and also mention them in the changelog.

File: MapboxNavigation/RouteMapViewController.swift
Comment (1ec5) [inline MapboxNavigation/RouteMapViewController.swift:18]: This type alias won’t be necessary if we declare the enumeration at the top level.

File: MapboxNavigation/NavigationViewController.swift
Comment (1ec5) [inline MapboxNavigation/NavigationViewController.swift:214]: Should we require these buttons to be of type FloatingButton so that custom buttons will look and behave consistently? If so, FloatingButton and `FloatingButton.rounded(image:selectedImage:size:)` should be documented and made public.

File: MapboxNavigation/NavigationViewController.swift
Comment (ShanMa1991) [reply MapboxNavigation/NavigationViewController.swift:203]: Thanks for the suggestion, I'll make it public at a top level to avoid typealias and nested type.

File: MapboxNavigation/NavigationView.swift
Comment (ShanMa1991) [reply MapboxNavigation/NavigationView.swift:78]: Thanks for the suggestion, I'll update the name and add it to the jazzy

File: MapboxNavigation/NavigationViewLayout.swift
Comment (ShanMa1991) [reply MapboxNavigation/NavigationViewLayout.swift:34]: Thanks for the suggestion, I'll update the case name and remove the default

File: MapboxNavigation/NavigationViewController.swift
Comment (ShanMa1991) [reply MapboxNavigation/NavigationViewController.swift:214]: Thanks for the suggestion. I'll keep it to be `UIButton` to be more versatile in development, but add the reference of the   `FloatingButton.rounded(image:selectedImage:size:)` in the documentation.

File: MapboxNavigation/NavigationViewController.swift
Comment (1ec5) [inline MapboxNavigation/NavigationViewController.swift:225]: We should make `FloatingButton.rounded(image:selectedImage:size:)` public and document it.

File: MapboxNavigation/NavigationViewController.swift
Comment (1ec5) [inline MapboxNavigation/NavigationViewController.swift:19]: Remember to add a documentation comment for both of these cases. Otherwise, the jazzy docset will show an enumeration with no cases.

**Review by 1ec5 — APPROVED** [review]
Good to merge once the FloatingButton method is public and the requested documentation is added.

File: MapboxNavigation/NavigationViewController.swift
Comment (ShanMa1991) [reply MapboxNavigation/NavigationViewController.swift:225]: Thanks for the suggestion! I'll work on this part.

File: MapboxNavigation/NavigationViewController.swift
Comment (ShanMa1991) [reply MapboxNavigation/NavigationViewController.swift:19]: Thanks for the suggestion, I'll add the two cases to documentation.

File: MapboxNavigation/Style.swift
Comment (1ec5) [inline MapboxNavigation/Style.swift:69]: > A rounded button with an icon that is designed to float above `NavigationMapView`.

File: MapboxNavigation/Style.swift
Comment (1ec5) [inline MapboxNavigation/Style.swift:108]: > The size of this button, or `FloatingButton.buttonSize` if this argument is not specified.

File: MapboxNavigation/Style.swift
Comment (1ec5) [inline MapboxNavigation/Style.swift:74]: > The default size of a floating button.

File: MapboxNavigation/NavigationViewController.swift
Comment (1ec5) [inline MapboxNavigation/NavigationViewController.swift:20]: > The top-left corner when using a left-to-right language or the top-right corner when using a right-to-left language.

File: MapboxNavigation/NavigationViewController.swift
Comment (1ec5) [inline MapboxNavigation/NavigationViewController.swift:24]: > The top-right corner when using a left-to-right language or the top-left corner when using a right-to-left language.

**Review by 1ec5 — APPROVED** [review]
Some copyediting nits, in case you have a chance to incorporate them before merging. Thanks for the quick turnaround on this feature!

File: MapboxNavigation/Style.swift
Comment (ShanMa1991) [reply MapboxNavigation/Style.swift]: Thanks for the suggestion!  I've edited the documentation as suggested.

**Pull request merged**
